pkg/envoy/resources.go

```diff
@@ -27,9 +37,92 @@ const (
 	NetworkPolicyHostsTypeURL = "type.googleapis.com/cilium.NetworkPolicyHosts"
 )
 
+// NPHDSCache is a cache of resources in the Network Policy Hosts Discovery
+// Service.
+type NPHDSCache struct {
+	*xds.Cache
+}
+
+func newNPHDSCache() NPHDSCache {
+	return NPHDSCache{Cache: xds.NewCache()}
+}
+
 var (
 	// NetworkPolicyHostsCache is the global cache of resources of type
 	// NetworkPolicyHosts. Resources in this cache must have the
 	// NetworkPolicyHostsTypeURL type URL.
-	NetworkPolicyHostsCache = xds.NewCache()
+	NetworkPolicyHostsCache = newNPHDSCache()
 )
+
+// OnIPIdentityCacheGC is required to implement IPIdentityMappingListener.
+func (cache *NPHDSCache) OnIPIdentityCacheGC() {
+	// We don't have anything to synchronize in this case.
+}
+
+// OnIPIdentityCacheChange pushes modifications to the IP<->Identity mapping
+// into the Network Policy Host Discovery Service (NPHDS).
+func (cache *NPHDSCache) OnIPIdentityCacheChange(
+	modType ipcache.CacheModification, ipIDPair identity.IPIdentityPair) {
+
+	// Look up the current resources for the specified Identity.
+	msg, err := cache.Lookup(NetworkPolicyHostsTypeURL, ipIDPair.ID.StringID())
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			logfields.IPAddr:       ipIDPair.IP,
+			logfields.Identity:     ipIDPair.ID,
+			logfields.Modification: modType,
+		}).Warningf("Can't update NPHDS cache")
+		return
+	}
+
+	if msg == nil && modType == ipcache.Delete {
+		// Doesn't exist; already deleted.
+		return
+	}
+
+	var npHost *envoyAPI.NetworkPolicyHosts
+	if msg == nil {
+		npHost = &envoyAPI.NetworkPolicyHosts{Policy: uint64(ipIDPair.ID), HostAddresses: make([]string, 0, 1)}
+	} else {
+		npHost = msg.(*envoyAPI.NetworkPolicyHosts)
+	}
+	if modType == ipcache.Upsert {
```

Comment:
I think this would cleaner as:

```
switch modType {
case ipcache.Upsert:
	if msg == nil {
		npHost = ...
	} else {
		npHost = msg.(*envoyAPI.NetworkPolicyHosts)
	}
	...
	cache.Upsert(...)
case ipcache.Delete:
	if msg == nil {
		// Doesn't exist; already deleted.
		return
	} else {
		cache.handleIPDelete(...)
	}
}
```